consensus/istanbul/backend/val_enode_share.go

```diff
@@ -0,0 +1,249 @@
+	i := 0
+	for address, validatorEnode := range sb.valEnodeTable.valEnodeTable {
+		sharedValidatorEnodes[i] = sharedValidatorEnode{
+			Address:  address,
+			EnodeURL: validatorEnode.enodeURL,
+			View:     validatorEnode.view,
+		}
+		i++
+	}
+
+	msg := &valEnodeShareMessage{
+		Address:   sb.Address(),
+		ValEnodes: sharedValidatorEnodes,
+	}
+
+	// Sign the validator enode share message
+	if err := msg.Sign(sb.Sign); err != nil {
+		sb.logger.Error("Error in signing an Istanbul Validator Enode Share message", "ValEnodeShareMsg", msg.String(), "err", err)
+		return nil, err
+	}
+
+	// Convert to payload
+	payload, err := msg.Payload()
+	if err != nil {
+		sb.logger.Error("Error in converting Istanbul Validator Enode Share message to payload", "ValEnodeShareMsg", msg.String(), "err", err)
+		return nil, err
+	}
+
+	sb.logger.Trace("Generated a Istanbul Validator Enode Share message", "ValEnodeShareMsg", msg.String())
+
+	return payload, nil
+}
+
+func (sb *Backend) sendValEnodeShareMsg() error {
+	payload, err := sb.generateValEnodeShareMsg()
+	if err != nil {
+		return err
+	}
+
+	if payload == nil {
+		return nil
+	}
+
+	if sb.broadcaster != nil {
+		sentryPeers := sb.broadcaster.GetSentryPeers()
+		if len(sentryPeers) > 0 {
+			sb.logger.Trace("Sending Istanbul Validator Enode Share payload to sentry peers", "sentry peer count", len(sentryPeers))
+			for _, sentryPeer := range sentryPeers {
+				sentryPeer.Send(istanbulValEnodeShareMsg, payload)
+			}
+		} else {
+			sb.logger.Warn("No sentry peers, cannot send Istanbul Validator Enode Share message")
+		}
+	}
+
+	return nil
+}
+
+// TODO: once we add a command line flag indicating a sentry is proxying for a
+// certain validator, add a check in here to make sure the message came from
+// the correct validator.
+func (sb *Backend) handleValEnodeShareMsg(payload []byte) error {
+	sb.logger.Warn("Handling an Istanbul Validator Enode message")
+
+	msg := new(valEnodeShareMessage)
+	// Decode message
+	err := msg.FromPayload(payload)
+	if err != nil {
+		sb.logger.Error("Error in decoding received Istanbul Validator Enode Share Message message", "err", err, "payload", hex.EncodeToString(payload))
+		return err
+	}
+
+	// Verify message signature
+	if err := msg.VerifySig(); err != nil {
+		sb.logger.Error("Error in verifying the signature of an Istanbul Validator Enode Share message", "err", err, "ValEnodeShareMsg", msg.String())
+		return err
+	}
+
+	sb.logger.Trace("Received an Istanbul Validator Enode Share message", "msg", msg.String())
+
+	block := sb.currentBlock()
+	valSet := sb.getValidators(block.Number().Uint64(), block.Hash())
+	for _, sharedValidatorEnode := range msg.ValEnodes {
+		valEnode := &validatorEnode{
+			enodeURL: sharedValidatorEnode.EnodeURL,
+			view:     sharedValidatorEnode.View,
+		}
+		if err := sb.valEnodeTable.upsert(sharedValidatorEnode.Address, valEnode, valSet, sb.Address(), true); err != nil {
```

Comment:
Everytime `upsert` is called, it will lock on the validator enode table mutex.

I think that it should lock once per `handleValEnodeShareMsg` invocation.   Also, this will handle the case in case two separate `handleValEnocdeShareMsg` invocations run concurrently (which would be rare, but not impossible).